### DISCUSSION TITLE
TerrainExtension with TileLayer

### DISCUSSION THREAD

**Discussion by radoslavirha** [visgl/deck.gl#7737] [Q&A]
Hi,

I'm playing little bit with TerrainExtension and I wanted to try it with TileLayers.

I don't think TerrainExtension is primarily meant to be used with TileLayer, but I thought it would be great.

Why:
When somebody wants to switch the texture on the terrain (pretty often as I found out). The solution was to have multiple TerrainLayers with different textures. It needs reloading elevation data so it's not very suitable.

Other way (not tested) would be just changing `texture` parameter and force reload .... I don't prefer this method, even if it would work.

The best way from my point of view is just single Terrain layer without texture and multiple separate TileLayers automatically rendering on the terrain.

Thanks

**Comment by ibgreen** [comment]
Thanks for the feedback. 

FWIW, a direction that we have been exploring is to externalize the tile data sources and not have those being part of (or strongly associated with) layers. Doing the same for terrain data sources would be a natural extension.

An initial aspect of this is an ongoing effort to break out the `Tileset2D` class from the `layers `module and break its strong dependency on deck.gl viewports. 

In the end state you would be able to create a single `Tileset2D` representing a texture source and the use it as the data prop for multiple tile layers. This common Tilese2D instance would reuse loaded tiles visible in multiple layers), and also allow an application to interact programmatically with a Tile data source without creating any layers. 

In the same vein we could allow a Tileset2D (or a derivative ElevationTileset) class to represent an elevation source. And we could assign those instances to the elevation property of layers, and switch them independently.

We could still support something like the current prop interface, if URLs are passed we would still create those Tileset2Ds under the hood.

This is part of a bigger ambition to separate loading and compute from rendering in layers to give applications more control and support more analytics use cases in addition to just rendering.

Does this align somewhat with what you are thinking?
